src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java

```diff
@@ -514,33 +528,62 @@ private boolean isSkipSuppressed(Request request) {
     HttpClientRequest httpClientRequest = buildInventoryQuery(httpClient, request);
     BatchStreamWrapper databaseWriteStream = getBatchHttpStream(httpClient, oaiPmhResponsePromise, httpClientRequest, vertxContext);
     httpClientRequest.sendHead();
+
+    ArrayDeque<Promise<Connection>> queue;
+    try {
+      queue =  (ArrayDeque<Promise<Connection>>)
+        getValueFrom(getValueFrom(getValueFrom(postgresClient, "client"), "pool"),
+          "waiters");
+    } catch (NullPointerException ex){
+      logger.error("Cannot get the pool size. Object is null.");
+      oaiPmhResponsePromise.fail(new IllegalArgumentException("Cannot get the pool size. Object is null."));
+      return completePromise;
+    }
+
+    databaseWriteStream.setCapacityChecker(()-> queue.size() > 20);
+
     databaseWriteStream.handleBatch(batch -> {
+      Promise<Void> savePromise = saveInstancesIds(batch, request, requestId, postgresClient, databaseWriteStream);
+      final Long returnedCount = databaseWriteStream.getReturnedCount();
 
-      Promise<Void> savePromise = saveInstancesIds(batch, request, requestId, postgresClient);
-      logger.info("Batch progress: " + batch.size() + "_" + databaseWriteStream.isStreamEnded()
-        + "_" + databaseWriteStream.getItemsInQueueCount() + "_" + databaseWriteStream.getReturnedCount());
+      if (returnedCount % 1000 == 0) {
+        logger.info("Batch saving progress: " + returnedCount + " returned so far, batch size: " + batch.size() + ", http ended: " + databaseWriteStream.isStreamEnded());
+      }
 
       if (databaseWriteStream.isTheLastBatch()) {
         savePromise.future().toCompletionStage().thenRun(completePromise::complete);
       }
+      databaseWriteStream.invokeDrainHandler();
     });
 
     return completePromise;
   }
 
+  private Object getValueFrom(Object obj, String fieldName) {
+    Field field = requireNonNull(ReflectionUtils.findField(requireNonNull(obj.getClass()), fieldName));
```

Comment:
Here the reflection is used by the reason that we need to have an access to vert.x "waiters" objects which are accumulated when batches are saved to database, the handling batches from inventory view is performed match faster versus saving to database. By this reason in some time we got a lot of waiters objects wich holds many of others and this leads to OutOfMemory.

Reply:
In future we can consider using static AtomicInteger to count the number of current db requests. It will be more readable in code, but less reliable because wouldn't take into account other requests.

Reply:
@IlliaDaliek, could you please add some comment in the source code as well.